modules/rgbd/src/hash_tsdf.cpp

```diff
@@ -317,19 +335,222 @@ inline Point3f HashTSDFVolumeCPU::getNormalVoxel(Point3f point) const
+    for (int i = 0; i < 8; i++)
+        neighbourCoords[i] *= voxelSize;
+
+    Point3f pointVec(point);
+    Point3f normal = Point3f(0, 0, 0);
+
+    Point3f c0_pointPrev = pointVec, c0_pointNext = pointVec;
+    Point3f c1_pointPrev = pointVec, c1_pointNext = pointVec;
+    Point3f c2_pointPrev = pointVec, c2_pointNext = pointVec;
+
+    c0_pointPrev.x -= voxelSize; c0_pointNext.x += voxelSize;
+    c1_pointPrev.y -= voxelSize; c1_pointNext.y += voxelSize;
+    c2_pointPrev.z -= voxelSize; c2_pointNext.z += voxelSize;
+
+    float c0_n_vx[8], c1_n_vx[8], c2_n_vx[8];
+    float c0_p_vx[8], c1_p_vx[8], c2_p_vx[8];
+
+    cv::Vec3i c0_n_vui[8], c0_p_vui[8];
+
+    for (int i = 0; i < 8; i++)
+        c0_n_vui[i] = volumeToVolumeUnitIdx(neighbourCoords[i] + c0_pointNext);
+    if (vuiCheck(c0_n_vui))
+    {
+        VolumeUnitMap::const_iterator it = volumeUnits.find(c0_n_vui[0]);
+        for (int i = 0; i < 8; i++)
+            c0_n_vx[i] = tsdfToFloat(_at_3arg(neighbourCoords[i] + c0_pointNext, c0_n_vui[0], it).tsdf);
+    }
+    else {
+        for (int i = 0; i < 8; i++)
+            c0_n_vx[i] = tsdfToFloat(_at_2arg(neighbourCoords[i] + c0_pointNext, c0_n_vui[i]).tsdf);
+    }
+
+    for (int i = 0; i < 8; i++)
+        c0_p_vui[i] = volumeToVolumeUnitIdx(neighbourCoords[i] + c0_pointPrev);
+    if (vuiCheck(c0_p_vui))
+    {
+        VolumeUnitMap::const_iterator it = volumeUnits.find(c0_p_vui[0]);
+        for (int i = 0; i < 8; i++)
+            c0_p_vx[i] = tsdfToFloat(_at_3arg(neighbourCoords[i] + c0_pointPrev, c0_p_vui[0], it).tsdf);
+    } else {
+        for (int i = 0; i < 8; i++)
+            c0_p_vx[i] = tsdfToFloat(_at_2arg(neighbourCoords[i] + c0_pointPrev, c0_p_vui[i]).tsdf);
+    }
+    for (int i = 0; i < 8; i++)
+        if (i % 4 > 1)
+            c1_n_vx[i] = tsdfToFloat(_at_1arg(neighbourCoords[i] + c1_pointNext).tsdf);
+        else
+            c1_p_vx[i] = tsdfToFloat(_at_1arg(neighbourCoords[i] + c1_pointPrev).tsdf);
+
+    for (int i = 0; i < 8; i++)
+        if (i % 2 == 1)
+            c2_n_vx[i] = tsdfToFloat(_at_1arg(neighbourCoords[i] + c2_pointNext).tsdf);
+        else
+            c2_p_vx[i] = tsdfToFloat(_at_1arg(neighbourCoords[i] + c2_pointPrev).tsdf);
+
+    c1_n_vx[0] = c0_p_vx[6];
+    c1_n_vx[1] = c0_p_vx[7];
+    c1_p_vx[2] = c0_n_vx[2];
+    c1_p_vx[3] = c0_n_vx[3];
+    c1_n_vx[4] = c0_p_vx[4];
+    c1_n_vx[5] = c0_p_vx[7];
+    c1_p_vx[6] = c0_n_vx[0];
+    c1_p_vx[7] = c0_n_vx[1];
+
+    c2_n_vx[0] = c0_p_vx[5];
+    c2_p_vx[1] = c0_p_vx[7];
+    c2_n_vx[2] = c0_n_vx[1];
+    c2_p_vx[3] = c0_n_vx[3];
+    c2_n_vx[4] = c0_p_vx[4];
+    c2_p_vx[5] = c0_p_vx[6];
+    c2_n_vx[6] = c0_n_vx[0];
+    c2_p_vx[7] = c0_n_vx[2];
+
+    normal.x = interpolate(c0_pointNext, c0_n_vx) - interpolate(c0_pointPrev, c0_p_vx);
+    normal.y = interpolate(c1_pointNext, c1_n_vx) - interpolate(c1_pointPrev, c1_p_vx);
+    normal.z = interpolate(c2_pointNext, c2_n_vx) - interpolate(c2_pointPrev, c2_p_vx);
```

Comment:
Here each subtraction of 2 `interpolate`-s can be shrunk to just 1 `interpolate`:
* that way is used in `tsdf.cpp` in `getNormalVoxel`, please refer there for details
* this works because `c{0,1,2}_pointNext` and `c{0,1,2}_pointPrev` differ by 1 in each dimension, therefore they share the same fractional part

Reply:
Like this:
```
float c0_vx[8], c1_vx[8], c2_vx[8];
for(int i = 0; i < 8; i++)
{
    c0_vx[i] = c0_n_vx[i] - c0_p_vx[i];
}
normal.x = interpolate(c0_pointNext, c0_vx);
...
	```

Reply:
In fact, all interpolation coeffs are the same at each interpolation

Reply:
done